### PR TITLE
NewNonceTransactionHandlerV2 with createAddressNonceHandlerFunc

### PR DESCRIPTION
**Pull request by dragos-rebegea**

### CODE REVIEW

File: interactors/nonceHandlerV2/export_test.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	erdgoCore "github.com/ElrondNetwork/elrond-sdk-erdgo/core"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/data"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/interactors"
+)
+
+// NewAddressNonceHandler returns a new instance of a addressNonceHandler
+func NewAddressNonceHandlerWithPrivateAccess(proxy interactors.Proxy, address erdgoCore.AddressHandler) (*addressNonceHandler, error) {
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV2/export_test.go:11]: comments mismatch

File: interactors/nonceHandlerV2/nonceTransactionsHandler.go [hunk 2]
```diff
@@ -17,6 +17,13 @@ const minimumIntervalToResend = time.Second
 
 var log = logger.GetOrCreate("elrond-sdk-erdgo/interactors/nonceHandlerV2")
 
+// ArgsNonceTransactionsHandlerV2 is the argument DTO for a pair
+type ArgsNonceTransactionsHandlerV2 struct {
+	Proxy                         interactors.Proxy
+	IntervalToResend              time.Duration
+	CreateAddressNonceHandlerFunc func(proxy interactors.Proxy, address core.AddressHandler) (interactors.AddressNonceHandler, error)
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV2/nonceTransactionsHandler.go:24]: Personally, I do not like this approach of providing this handler function
The better way to do this is to apply the abstract factory pattern:
1. create 2 small structures that implement the same interface "AddressNonceHandlerCreator", each defining the following method:
```GO
Create(proxy interactors.Proxy, address core.AddressHandler) (interactors.AddressNonceHandler, error)
```
2. use the interface in this Args struct, calling the Create method where needed
3. create unit tests for all implementations. Surprisingly, those structs can have 100% code coverage

File: interactors/nonceHandlerV2/export_test.go (hunk 1 above)
Comment (dragos-rebegea) [reply interactors/nonceHandlerV2/export_test.go:11]: done

File: interactors/nonceHandlerV2/nonceTransactionsHandler.go [hunk 3]
```diff
@@ -28,30 +35,35 @@ var log = logger.GetOrCreate("elrond-sdk-erdgo/interactors/nonceHandlerV2")
...
+		proxy:            args.Proxy,
+		handlers:         make(map[string]interactors.AddressNonceHandler),
+		intervalToResend: args.IntervalToResend,
+		creator:          args.Creator,
 	}
 
 	ctx, cancelFunc := context.WithCancel(context.Background())
 	nth.cancelFunc = cancelFunc
-	go nth.resendTransactionsLoop(ctx, intervalToResend)
+	go nth.resendTransactionsLoop(ctx, args.IntervalToResend)
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV2/nonceTransactionsHandler.go:66]: useless passing of the interval argument, on L152 & L160 could have used nth.intervalToResend

File: interactors/nonceHandlerV2/nonceTransactionsHandler_test.go [hunk 4]
```diff
@@ -22,30 +23,65 @@ var txArgs = data.ArgCreateTransaction{}
...
-	require.Nil(t, nth)
-	assert.True(t, errors.Is(err, interactors.ErrInvalidValue))
-	assert.True(t, strings.Contains(err.Error(), "for intervalToResend in NewNonceTransactionHandlerV2"))
+		args := createMockArgsNonceTransactionsHandlerV2()
+		args.Proxy = nil
+		nth, err := NewNonceTransactionHandlerV2(args)
+		require.Nil(t, nth)
+		assert.Equal(t, interactors.ErrNilProxy, err)
+	})
+	t.Run("nil proxy", func(t *testing.T) {
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV2/nonceTransactionsHandler_test.go:35]: wrong test name + duplicated test with L45

File: interactors/nonceHandlerV2/nonceTransactionsHandler_test.go [hunk 5]
```diff
@@ -22,30 +23,65 @@ var txArgs = data.ArgCreateTransaction{}
...
+		t.Parallel()
+
+		args := createMockArgsNonceTransactionsHandlerV2()
+		args.IntervalToResend = time.Second - time.Nanosecond
+		nth, err := NewNonceTransactionHandlerV2(args)
+		require.Nil(t, nth)
+		assert.True(t, errors.Is(err, interactors.ErrInvalidValue))
+		assert.True(t, strings.Contains(err.Error(), "for intervalToResend in NewNonceTransactionHandlerV2"))
+	})
+	t.Run("nil func handler", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (dragos-rebegea) [reply interactors/nonceHandlerV2/nonceTransactionsHandler_test.go:55]: done

File: interactors/nonceHandlerV2/addressNonceHandlerCreator.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package nonceHandlerV2
+
+import (
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/core"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/interactors"
+)
+
+//AddressNonceHandlerCreator is used to create addressNonceHandler instances
```
Comment (sstanculeanu) [inline interactors/nonceHandlerV2/addressNonceHandlerCreator.go:8]:
```suggestion
// AddressNonceHandlerCreator is used to create addressNonceHandler instances
```

File: interactors/nonceHandlerV2/addressNonceHandlerCreator_test.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+	"crypto/rand"
+	"fmt"
+	"testing"
+
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/data"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/testsCommon"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAddressNonceHandlerCreator_Create(t *testing.T) {
```
Comment (sstanculeanu) [inline interactors/nonceHandlerV2/addressNonceHandlerCreator_test.go:13]: `t.Parallel()`

File: interactors/nonceHandlerV2/nonceTransactionsHandler.go [hunk 8]
```diff
@@ -17,6 +17,13 @@ const minimumIntervalToResend = time.Second
 
 var log = logger.GetOrCreate("elrond-sdk-erdgo/interactors/nonceHandlerV2")
 
+// ArgsNonceTransactionsHandlerV2 is the argument DTO for a pair
```
Comment (sstanculeanu) [inline interactors/nonceHandlerV2/nonceTransactionsHandler.go:20]: incomplete comment?

File: interactors/nonceHandlerV2/singleTransactionAddressNonceHandlerCreator.go [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package nonceHandlerV2
+
+import (
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/core"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/interactors"
+)
+
+//SingleTransactionAddressNonceHandlerCreator is used to create singleTransactionAddressNonceHandler instances
```
Comment (sstanculeanu) [inline interactors/nonceHandlerV2/singleTransactionAddressNonceHandlerCreator.go:8]:
```suggestion
// SingleTransactionAddressNonceHandlerCreator is used to create singleTransactionAddressNonceHandler instances
```

File: interactors/nonceHandlerV2/singleTransactionAddressNonceHandlerCreator.go [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+type SingleTransactionAddressNonceHandlerCreator struct{}
```
Comment (sstanculeanu) [inline interactors/nonceHandlerV2/singleTransactionAddressNonceHandlerCreator.go:9]: no need to be exported

File: interactors/nonceHandlerV2/addressNonceHandlerCreator.go [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+type AddressNonceHandlerCreator struct{}
```
Comment (sstanculeanu) [inline interactors/nonceHandlerV2/addressNonceHandlerCreator.go:9]: no need to be exported

File: interactors/nonceHandlerV2/singleTransactionAddressNonceHandlerCreator_test.go [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+	"crypto/rand"
+	"fmt"
+	"testing"
+
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/data"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/testsCommon"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSingleTransactionAddressNonceHandlerCreator_Create(t *testing.T) {
```
Comment (sstanculeanu) [inline interactors/nonceHandlerV2/singleTransactionAddressNonceHandlerCreator_test.go:13]: `t.Parallel()`

File: interactors/nonceHandlerV2/addressNonceHandlerCreator.go (hunk 6 above)
Comment (dragos-rebegea) [reply interactors/nonceHandlerV2/addressNonceHandlerCreator.go:8]: done

File: interactors/nonceHandlerV2/nonceTransactionsHandler.go (hunk 8 above)
Comment (dragos-rebegea) [reply interactors/nonceHandlerV2/nonceTransactionsHandler.go:20]: done

File: interactors/nonceHandlerV2/addressNonceHandlerCreator_test.go (hunk 7 above)
Comment (dragos-rebegea) [reply interactors/nonceHandlerV2/addressNonceHandlerCreator_test.go:13]: done

File: interactors/nonceHandlerV2/singleTransactionAddressNonceHandlerCreator.go (hunk 9 above)
Comment (dragos-rebegea) [reply interactors/nonceHandlerV2/singleTransactionAddressNonceHandlerCreator.go:8]: done

File: interactors/nonceHandlerV2/singleTransactionAddressNonceHandlerCreator.go (hunk 10 above)
Comment (dragos-rebegea) [reply interactors/nonceHandlerV2/singleTransactionAddressNonceHandlerCreator.go:9]: done

File: interactors/nonceHandlerV2/addressNonceHandlerCreator.go (hunk 11 above)
Comment (dragos-rebegea) [reply interactors/nonceHandlerV2/addressNonceHandlerCreator.go:9]: done

File: interactors/nonceHandlerV2/singleTransactionAddressNonceHandlerCreator_test.go (hunk 12 above)
Comment (dragos-rebegea) [reply interactors/nonceHandlerV2/singleTransactionAddressNonceHandlerCreator_test.go:13]: done

File: interactors/nonceHandlerV2/addressNonceHandlerCreator.go (hunk 11 above)
Comment (sstanculeanu) [reply interactors/nonceHandlerV2/addressNonceHandlerCreator.go:9]: still exported

**Review by iulianpascalau — APPROVED** [review]

**Review by sstanculeanu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**